src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -357,7 +357,13 @@ infer' (Abs binder ret)
       ty <- freshType
       withBindingGroupVisible $ bindLocalVariables [(arg, ty, Defined)] $ do
         body@(TypedValue _ _ bodyTy) <- infer' ret
-        return $ TypedValue True (Abs (VarBinder arg) body) $ function ty bodyTy
+        let (qs, cs, ty') = moveQuantifiersToFrontOverFunction (function ty bodyTy)
```

Comment:
Can you not just instantiate the type of the body here?

Reply:
Yes, that makes more sense. 😸